### ISSUE TITLE
PermitIf == false should not trigger OnUnhandledTrigger exception.

### ISSUE THREAD

**Issue by TybaIt** [dotnet-state-machine/stateless#412]
Apparently, when the condition in ``PermitIf`` returns false and does not allow transition it throws an OnUnhandledTrigger exception.

you can see this in my code here:

https://github.com/TybaIt/Blish-HUD-Modules/blob/5d1265dc5519263263bccd9b6f3c993cd67bb129/Music%20Mixer%20Module/Gw2StateService.cs#L153
The logs were getting spammed when the ``_toggle`` was false (and thus PermitIf didn't allow transition) by doing a switch case in the override of ``OnUnhandledTrigger``

**Comment by HenningNT** [comment]
I agree, this is annoying.

If I find the time I might change the behaviour, but not in the near future.

**Comment by DeepakParamkusam** [comment]
I disagree.

Correct me if I am wrong, but how else can we know during runtime that a valid trigger was received but a guard condition failed? We use this feature extensively in our application and silently failing is not an option.

**Comment by HenningNT** [comment]
Of course, there are many use cases, and many ways to use the library. If it's an requirement to keep track of failed guard condition, then the OnUnhandledTrigger event is very convenient.

As a compromise, we could maybe add a OnGuardFailed in addition to OnUnhandledTrigger? One would fire if there are no transitions at all, and the other if there are no valid (i.e. all guards fail) transitions available.

But I don't see this as the most important matter right now...

**Comment by DeepakParamkusam** [comment]
That would be nice, but it would be a breaking change and I'm always wary of those.

**Comment by TybaIt** [comment]
I must say OnUnhandledTriggerException should be thrown **when a trigger is not defined** but PermitIf does define the trigger, but it isn't permitted during runtime. So it makes no sense at all according to the doc that it throws OnUnhandledTriggerException.

Please stop letting it throw when a trigger is defined and a guard just returns false on runtime.

The compromisse is a great idea but it does not solve my initial issue with the throw definition of said exception.

**Comment by HenningNT** [comment]
I agree, the name of the exception is somewhat misleading, when the trigger has been "handled", but the guard function prevents the transition from completing.
